internal/executor/job/job_run_state_store.go

```diff
@@ -81,25 +95,59 @@ func (stateStore *JobRunStateStore) reportRunActive(pod *v1.Pod) {
 
 	currentState.Phase = Active
 	currentState.KubernetesId = string(pod.UID)
-	currentState.LastTransitionTime = time.Now()
+	currentState.Job = nil // Now that the job is active, remove the object to save memory
+	currentState.LastPhaseTransitionTime = time.Now()
 }
 
-func (stateStore *JobRunStateStore) ReportRunLeased(runMeta *RunMeta) {
+func (stateStore *JobRunStateStore) ReportRunLeased(runMeta *RunMeta, job *SubmitJob) {
 	stateStore.lock.Lock()
 	defer stateStore.lock.Unlock()
 	_, present := stateStore.jobRunState[runMeta.RunId]
 	if !present {
 		state := &RunState{
-			Meta:               runMeta,
-			Phase:              Leased,
-			LastTransitionTime: time.Now(),
+			Meta:                    runMeta,
+			Job:                     job,
+			Phase:                   Leased,
+			LastPhaseTransitionTime: time.Now(),
 		}
 		stateStore.jobRunState[runMeta.RunId] = state
 	} else {
 		log.Warnf("run unexpectedly reported as leased (runId=%s, jobId=%s), state already exists", runMeta.RunId, runMeta.JobId)
 	}
 }
 
+func (stateStore *JobRunStateStore) ReportRunInvalid(runMeta *RunMeta) {
+	stateStore.lock.Lock()
+	defer stateStore.lock.Unlock()
+	_, present := stateStore.jobRunState[runMeta.RunId]
+	if !present {
+		state := &RunState{
+			Meta:                    runMeta,
+			Phase:                   Invalid,
+			LastPhaseTransitionTime: time.Now(),
+		}
+		stateStore.jobRunState[runMeta.RunId] = state
+	} else {
+		log.Warnf("run unexpectedly reported as invalid (runId=%s, jobId=%s), state already exists", runMeta.RunId, runMeta.JobId)
+	}
+}
+
+func (stateStore *JobRunStateStore) ReportSuccessfulSubmission(runMeta *RunMeta) {
+	stateStore.lock.Lock()
+	defer stateStore.lock.Unlock()
+
+	currentState, present := stateStore.jobRunState[runMeta.RunId]
+	if !present {
+		log.Warnf("run unexpected reported as failed submission (runId=%s, jobId=%s), no current state exists", runMeta.RunId, runMeta.JobId)
+		currentState = &RunState{
+			Meta: runMeta,
+		}
+		stateStore.jobRunState[runMeta.RunId] = currentState
```

Comment:
This should never be possible and should just warn + exit. Let me fix it

Reply:
Actually looking at it - it would be completely fine. However it is handling a case that should never happen - so probably fine to just do as I said above